`pnpm publish --json` does not report error

### pnpm version:

7.17.1

### Code to reproduce the issue:

It's about publish so hard to create a repo.

The basic idea is that when there are error, e.g.:

```sh
pnpm publish

 ERR_PNPM_CANNOT_RESOLVE_WORKSPACE_PROTOCOL  Cannot resolve workspace protocol of dependency "@mocktomata/plugin-fixture-esm-b" because this dependency is not installed. Try running "pnpm install".
```

Running it with `pnpm publish --json` does not report such error.
It only returns code = 1


### Expected behavior:

The json should contains the error causing the failure

### Actual behavior:

No json returned

### Additional information:

 - `node -v` prints: 16.18.0
 - Windows, macOS, or Linux?: Windows and Linux

This is the root cause of: https://github.com/changesets/changesets/issues/1003

--- a/.changeset/sharp-masks-leave.md
+++ b/.changeset/sharp-masks-leave.md
@@ -1,0 +1,5 @@
+---
+"pnpm": patch
+---
+
+The fatal error should be printed in JSON format, when running a pnpm command with the `--json` option [#5710](https://github.com/pnpm/pnpm/issues/5710).

--- a/pnpm/src/errorHandler.ts
+++ b/pnpm/src/errorHandler.ts
@@ -1,22 +1,29 @@
 import { logger } from '@pnpm/logger'
-import { REPORTER_INITIALIZED, printError } from './main'
+import { REPORTER_INITIALIZED } from './main'
 
-export function errorHandler (error: Error) {
-  if (!global[REPORTER_INITIALIZED]) {
-    printError(error.message)
-    process.exitCode = 1
-    return
-  }
-  if (global[REPORTER_INITIALIZED] === 'silent') {
-    process.exitCode = 1
-    return
-  }
+export function errorHandler (error: Error & { code?: string }) {
   if (error.name != null && error.name !== 'pnpm' && !error.name.startsWith('pnpm:')) {
     try {
       error.name = 'pnpm'
     } catch {
       // Sometimes the name property is read-only
     }
+  }
+
+  if (!global[REPORTER_INITIALIZED]) {
+    // print parseable error on unhandled exception
+    console.log(JSON.stringify({
+      error: {
+        code: error.code ?? error.name,
+        message: error.message,
+      },
+    }, null, 2))
+    process.exitCode = 1
+    return
+  }
+  if (global[REPORTER_INITIALIZED] === 'silent') {
+    process.exitCode = 1
+    return
   }
 
   // bole passes only the name, message and stack of an error

--- a/pnpm/src/errorHandler.ts
+++ b/pnpm/src/errorHandler.ts
@@ -1,9 +1,9 @@
 import { logger } from '@pnpm/logger'
-import { REPORTER_INITIALIZED } from './main'
+import { REPORTER_INITIALIZED, printError } from './main'
 
 export function errorHandler (error: Error) {
   if (!global[REPORTER_INITIALIZED]) {
-    console.log(error)
+    printError(error.message)
     process.exitCode = 1
     return
   }

--- a/pnpm/src/main.ts
+++ b/pnpm/src/main.ts
@@ -138,8 +138,8 @@ export async function main (inputArgv: string[]) {
       cmd,
       config,
     })
+    global[REPORTER_INITIALIZED] = reporterType
   }
-  global[REPORTER_INITIALIZED] = reporterType
 
   const selfUpdate = config.global && (cmd === 'add' || cmd === 'update') && cliParams.includes(packageManager.name)
 
@@ -296,7 +296,7 @@ export async function main (inputArgv: string[]) {
   }
 }
 
-function printError (message: string, hint?: string) {
+export function printError (message: string, hint?: string) {
   console.log(`${chalk.bgRed.black('\u2009ERROR\u2009')} ${chalk.red(message)}`)
   if (hint) {
     console.log(hint)

--- a/pnpm/src/main.ts
+++ b/pnpm/src/main.ts
@@ -296,7 +296,7 @@ export async function main (inputArgv: string[]) {
   }
 }
 
-export function printError (message: string, hint?: string) {
+function printError (message: string, hint?: string) {
   console.log(`${chalk.bgRed.black('\u2009ERROR\u2009')} ${chalk.red(message)}`)
   if (hint) {
     console.log(hint)

--- a/pnpm/test/errorHandler.test.ts
+++ b/pnpm/test/errorHandler.test.ts
@@ -1,0 +1,37 @@
+import { prepare, preparePackages } from '@pnpm/prepare'
+import writeYamlFile from 'write-yaml-file'
+import { execPnpmSync } from './utils'
+
+test('should print json format error when publish --json failed', async () => {
+  prepare({
+    name: 'test-publish-package-no-version',
+    version: undefined,
+  })
+
+  const { status, stdout } = execPnpmSync(['publish', '--dry-run', '--json'])
+
+  expect(status).toBe(1)
+  const { error } = JSON.parse(stdout.toString())
+  expect(error?.code).toBe('ERR_PNPM_PACKAGE_VERSION_NOT_FOUND')
+  expect(error?.message).toBe('Package version is not defined in the package.json.')
+})
+
+test('should print json format error when add dependency on workspace root', async () => {
+  preparePackages([
+    {
+      name: 'project-a',
+      version: '1.0.0',
+    },
+    {
+      name: 'project-b',
+      version: '1.0.0',
+    },
+  ])
+  await writeYamlFile('pnpm-workspace.yaml', { packages: ['**', '!store/**'] })
+
+  const { status, stdout } = execPnpmSync(['add', 'nanoid', '-p'])
+
+  expect(status).toBe(1)
+  const { error } = JSON.parse(stdout.toString())
+  expect(error?.code).toBe('ERR_PNPM_ADDING_TO_ROOT')
+})

